core/src/main/kotlin/com/strumenta/kolasu/serialization/JsonGenerator.kt

```diff
@@ -136,6 +164,17 @@ class JsonGenerator {
         }
     }
 
+    private fun computeIds(root: Node): IdentityHashMap<Node, String> =
+        IdentityHashMap<Node, String>().apply {
+            root.walk().forEach { this[it] = UUID.randomUUID().toString() }
```

Comment:
I would consider assigning IDs only to nodes which needs them as someone is referring to them. In this way it will make easier to find the referred nodes

Reply:
In order to be able to do this, `ReferenceByName` should be modelled as `NodeType` as the current walkers cannot retrieve these otherwise. Am I correct?

Reply:
Another possibility would be to process the properties, as follows:
```
val ids = node.computeIds {
    it.walk()
        .find { node -> node.properties
            .any { property -> property.value is ReferenceByName<*>
                     && (property.value as ReferenceByName<*>).referred == it}}  
        ?.let { "some_identifier" }
}
```

Reply:
`OnlyReferencedIdProvider` serves this purpose, it internally uses `SequentialIdProvider` by default but can be easily paremeterized.

Reply:
I think the alternative approach (which you implement) is better as the ReferenceByName does not represent a node per se, but a reference to a node